### PR TITLE
Python tests: skip importing weirdly-named modules

### PR DESCRIPTION
**Pull request by iarspider**
e.g. with dashes in name

### CODE REVIEW

**Comment by adamjstewart** [comment]
Can you give an example of such a file? I thought Python libraries had to use the same filename structure as variable (letters, numbers, underscore, but no dashes). Maybe we should just skip such files?

**Comment by iarspider** [comment]
@adamjstewart see modules for `py-jupyter-server`

**Comment by adamjstewart** [comment]
I have `py-jupyter-server@1.9.0` installed and I don't see any dashes. Did that change in newer releases? I'm not sure if this library is something that people import.

**Comment by iarspider** [comment]
@adamjstewart try 1.11.2:
```
$ spack install --test=root py-jupyter-server@1.11.2
(...)

/root/spack/lib/spack/spack/package.py:1893, in run_test:
       1890                if spack.config.get('config:fail_fast', False):
       1891                    raise TestFailure([(exc, m)])
       1892                else:
  >>   1893                    self.test_failures.append((exc, m))
       1894                return False

See build log for details:
  /tmp/root/spack-stage/spack-stage-py-jupyter-server-1.11.2-azzmxxhvo3dbaf6sdhgr7b7dlz3q745y/spack-build-out.txt


$ cat /tmp/root/spack-stage/spack-stage-py-jupyter-server-1.11.2-azzmxxhvo3dbaf6sdhgr7b7dlz3q745y/spack-build-out.txt
(...)
==> [2021-11-02-17:44:51.494804] checking import of jupyter_server.tests.namespace-package-test.test_namespace.test_package
==> [2021-11-02-17:44:51.495209] '/root/spack/opt/spack/linux-ubuntu20.04-skylake_avx512/gcc-9.3.0/python-3.8.12-lojmxwwgwen5ryic7glxiqrynlbd2xkr/bin/python3.8' '-c' 'import jupyter_server.tests.namespace-package-test.test_namespace.test_package'
  File "<string>", line 1
    import jupyter_server.tests.namespace-package-test.test_namespace.test_package
                                         ^
SyntaxError: invalid syntax
FAILED: Command exited with status 1:
    '/root/spack/opt/spack/linux-ubuntu20.04-skylake_avx512/gcc-9.3.0/python-3.8.12-lojmxwwgwen5ryic7glxiqrynlbd2xkr/bin/python3.8' '-c' 'import jupyter_server.tests.namespace-package-test.test_namespace.test_package'
  File "<string>", line 1
    import jupyter_server.tests.namespace-package-test.test_namespace.test_package
                                         ^
SyntaxError: invalid syntax
```

**Comment by adamjstewart** [comment]
Yeah, I would just skip that, I don't think it would import successfully anyway, at least for older Python versions.

Currently we check for `*.py` files in the site_packages dir and recursively check for `__init__.py` files. A better strategy would be to only traverse directories that contain a `__init__.py` file, but that's more complicated. Personally I would just skip any directory containing dashes in the name.

**Comment by iarspider** [comment]
@adamjstewart that's also an option. In general though, we may want to filter out anything that is not `[a-zA-Z0-9._]`, just in case some package will have a module with e.g. `%` or `+` in name.

**Comment by iarspider** [comment]
@adamjstewart gentle ping. The failure is not related to this MR.

File: lib/spack/spack/build_systems/python.py [hunk 1]
```diff
@@ -319,6 +320,9 @@ def test(self):
         # Make sure we are importing the installed modules,
         # not the ones in the source directory
         for module in self.import_modules:
+            if not re.match('[a-zA-Z0-9._]+$', module):
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/python.py]: Can we instead move this logic into `self.import_modules`?

Comment (iarspider) [reply lib/spack/spack/build_systems/python.py]: Sure.

File: lib/spack/spack/build_systems/python.py [hunk 2]
```diff
@@ -135,14 +136,20 @@ def import_modules(self):
...
         # Some Python libraries are modules: individual *.py files
         # found in the site-packages directory
         for path in find(root, '*.py', recursive=False):
-            modules.append(path.replace(root + os.sep, '', 1).replace(
-                '.py', '').replace('/', '.'))
+            mod = path.replace(root + os.sep, '', 1).replace(
+                '.py', '').replace('/', '.')
+            if not re.match('[a-zA-Z0-9._]+$', mod):
+                continue
+            modules.append(mod)
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/python.py]: There would be less code duplication if you append all modules to the list and then filter afterwards. I would personally use Python's `filter()` function with a lambda.

**Comment by iarspider** [comment]
@adamjstewart sure.

File: lib/spack/spack/build_systems/python.py (hunk 2 above)
Comment (adamjstewart) [reply lib/spack/spack/build_systems/python.py]: Still waiting on this change

Comment (iarspider) [reply lib/spack/spack/build_systems/python.py]: Sorry, forgot to push.

File: lib/spack/spack/build_systems/python.py [hunk 3]
```diff
@@ -135,14 +136,18 @@ def import_modules(self):
         # Some Python libraries are packages: collections of modules
         # distributed in directories containing __init__.py files
         for path in find(root, '__init__.py', recursive=True):
-            modules.append(path.replace(root + os.sep, '', 1).replace(
-                os.sep + '__init__.py', '').replace('/', '.'))
+            mod = path.replace(root + os.sep, '', 1).replace(
+                os.sep + '__init__.py', '').replace('/', '.')
+            modules.append(mod)
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/python.py]: You can undo this change to minimize the diff

**Review by adamjstewart — APPROVED** [review]
